Update EP7 vector data (#131)
carpentries-incubator/geospatial-python: _episodes/07-vector-data-in-python.md

~~~
{: .language-python}

Using the `bbox` input argument, we can load only the spatial features intersecting the provided bounding box.

~~~
# Partially load data within the bounding box
fields = gpd.read_file("data/brpgewaspercelen_definitief_2020_small.gpkg", bbox=bbox)
~~~
{: .language-python}

> ## How should I define my bounding box?
> For simplicity, here we assume the **Coordinate Reference System (CRS)** and **extent** of the vector file are known (for instance they are provided in the dataset documentation). 
> 

> > You can also define your bounding box with online coordinates visualization tools. For example, in our case, we can use the [RD-viewer](https://openstate.github.io/rdnewviewer/).

Reviewer: reading the changes further, I see why you selected this particular coordinate viewer, it looks like it is specific to this Dutch projection. I think it would still be useful to show a coordinate viewer like geojson.io since users will usually be needing to draw AOIs with lat lon coordinates rather than use RD Viewer. Can we mention both at make the distinction that most coordinate viewers use EPSG 4326 and lat/lon, and that RD Viewer in particular uses the RD projection?

> > You can also define your bounding box with online coordinates visualization tools. For example, in our case, we can use the [RD-viewer](https://openstate.github.io/rdnewviewer/).

Author: Thanks Ryan! I think it's a good idea. Living in the Netherlands does make me forget the international standards a bit. Will change this

> > You can also define your bounding box with online coordinates visualization tools. For example, in our case, we can use the [RD-viewer](https://openstate.github.io/rdnewviewer/).

Author: @rbavery I used your suggestion, kept the geojson.io with WGS84 as a default, and mentioning since we are working with a Dutch dataset, we can use the RD viewer.

> 

> > Some Python tools, e.g. [`fiona`](https://fiona.readthedocs.io/en/latest/)(which is also the backend of `geopandas`), provide the file inspection functionality without actually the need to read the full data set into memory. An example can be found in [the documentation of fiona](https://fiona.readthedocs.io/en/latest/manual.html#format-drivers-crs-bounds-and-schema).

Reviewer: ```suggestion
> Some Python tools, e.g. [`fiona`](https://fiona.readthedocs.io/en/latest/)(which is also the backend of `geopandas`), provide the file inspection functionality without the need to read the full data set into memory. An example can be found in [the documentation of fiona](https://fiona.readthedocs.io/en/latest/manual.html#format-drivers-crs-bounds-and-schema).
```

{: .callout}

And we can plot the overview by:
~~~
fields.plot()
~~~
{: .language-python}
![Cropped fields plot image](../fig/E07-02-fields.png)
{: .output}
## Vector Metadata & Attributes
When we read the vector dataset with Python (as our `fields` variable) it is loaded as a `GeoDataFrame` object. The `read_file()` function also automatically stores geospatial information about the data. We are particularly interested in describing the format, CRS, extent, and other components of the vector data, and the attributes which describe properties associated
with each vector object. 

> For example, we will explore

Reviewer: ```suggestion
We will explore
```

From now on, we will continue with the cropped fields data `fields_cx`. We will save it to a shapefile (`.shp`) and use it later. The `to_file` function can be used for exportation:

~~~
fields_cx.to_file('data/fields_cropped.shp')
~~~
{: .language-python}

This will write it to disk (in this case, in 'shapefile' format), containing only the data from our cropped area. It can be read again at a later time using the `read_file()` method we have been using above. Note that this actually writes multiple files to disk (`fields_cropped.cpg`, `fields_cropped.dbf`, `fields_cropped.prj`, `fields_cropped.shp`, `fields_cropped.shx`). All these files should ideally be present in order to re-read the dataset later, although only the `.shp`, `.shx`, and `.dbf` files are mandatory (see the [Introduction to Vector Data]({{site.baseurl}}/02-intro-to-vector-data) lesson for more information.)

## Selecting spatial features

From now on, we will take in a point dataset `brogmwvolledigeset.zip`, which is the underground water monitoring wells. We will perform vector processing on this dataset, together with the cropped field polygons `fields_cropped.shp`.

Let's read the two datasets.

~~~

> fields =  gpd.read_file("data/fields_cropped.shp")

Reviewer 2: Why re-reading the polygons?

Author: you are right, this comes from legacy, will remove.

wells = gpd.read_file("data/brogmwvolledigeset.zip")
~~~
{: .language-python}

And take a look at the wells:

~~~
wells.plot(markersize=0.1)
~~~
{: .language-python}
![All wells](../fig/E07-03-wells-nl.png)

The points represents all the wells over the Netherlands. Since the wells are in the lat/lon coordinates. To make it comparable with fields, we need to transfer the CRS to RD first:

~~~
wells = wells.to_crs(28992)
~~~
{: .language-python}

Now we would like to compare the wells with the cropped fields. We can select the wells within the fields using the `.clip` function:

~~~
wells_clip = wells.clip(fields)
wells_clip
~~~
{: .language-python}

~~~
bro_id delivery_accountable_party quality_regime  ...
40744  GMW000000043703  27364178    IMBRO/A    ...
38728  GMW000000045818  27364178    IMBRO/A    ...
...                ...       ...        ...    ...
40174  GMW000000043963  27364178    IMBRO/A    ...
19445  GMW000000024992  50200097    IMBRO/A    ...
[79 rows x 40 columns]
~~~
{: .output}

After this selection, all the wells outside fields are dropped. This takes a while to execute, because we are clipping a relatively large number of points with many polygons.

> 

Reviewer 2: I like this addition! But it might be easier to follow if you mention here where this is going (i.e. we first increase the size of the polygons such that they overlap and then we merge them)?

Author: Thx! I added some descriptions.

> ## Challenge: select fields within 500m from the wells
> This time, we will do a selection the other way around. Can you find and plot the field polygons (stored in `fields_cropped.shp`), which intersect the 500m radius of any wells (stored in `brogmwvolledigeset.zip`)? 
> 
> - Hint1: `brogmwvolledigeset.zip` is in CRS 4326. Don't forget the CRS conversion.
>
> - Hint2: `brogmwvolledigeset.zip` is big. To improve the performance, you may want to crop it first.
> 
> > ## Answers
> > ~~~
> > # Read in data
> > fields =  gpd.read_file("data/fields_cropped.shp")
> > wells = gpd.read_file("data/brogmwvolledigeset.zip")
> > 
> > # Crop points with bounding box
> > xmin, ymin, xmax, ymax = fields.total_bounds
> > wells = wells.to_crs(28992)
> > wells_cx = wells.cx[xmin-500:xmax+500, ymin-500:ymax+500]
> > 
> > # Create buffer
> > wells_cx_500mbuffer = wells_cx.copy()
> > wells_cx_500mbuffer['geometry'] = wells_cx.buffer(500)
> > 
> > # Clip
> > fields_clip_buffer = fields.clip(wells_cx_500mbuffer)
> > fields_clip_buffer.plot()
> > ~~~
> > {: .language-python}
> {: .solution}
{: .challenge}

## Spatially join the features

> Sometimes, we would like to search for an attribute in attribute in another `GeoDataFrame`, and add the results to the existing `GeoDataFrame`. For example, we may want to store the well IDs as a new column in the `fields` object. To do this, we can use the `sjoin` function to join two `GeoDataFrame`:

Reviewer 2: ```suggestion
Sometimes, we would like to match elements between two `GeoDataFrame`'s on the basis of their spatial relationship. For example, we may want to pair water wells to the crop fields they fall within. To do this, we can use the `sjoin` function to "spatially" join two `GeoDataFrame`'s:
```

~~~
{: .output}

This will result in a `GeodataFrame` of all polygons intersecting the wells, adding the well IDs as a new column: `bro_id`.

> ## Challenge: count the number of field polygons within a distance from the points
> Use the `sjoin` function to join two polygons: the field polygons (`fields`, read in from `data/fields_cropped.shp`) and the wells 500m buffer polygons (`wells_cx_500mbuffer` from the previous exercise). And answer the following questions: 
> 
> - How many fields are intersecting the 500m buffer of wells?
>
> - How many wells are intersecting the fields with a 500 buffer?
> 
> > ## Answers
> > ~~~
> > fields_wells_buffer = fields.sjoin(wells_cx_500mbuffer)
> > print(fields_wells_buffer.shape)
> > print(fields_wells_buffer.index.unique().shape) # Take the unique of the index
> > print(fields_wells_buffer['bro_id'].unique().shape) # Take the unique of the bro_ids
> > 
> > # validation
> > print(fields.clip(wells_cx_500mbuffer).shape)
> > ~~~
> > {: .language-python}
> > Note that a polygon can fall in the 500m buffer of multiple points. So we should use `.unique()` function to get the actual number of the polygons.
> {: .solution}
{: .challenge}

> ## (optional) Modify the geometry of a GeoDataFrame

Reviewer 2: Maybe just leave out the optional and every instructor can decide whether this is relevant or not for the audience?

```suggestion
## Modify the geometry of a GeoDataFrame
```

Author: Good suggestion, I will add it to the instrutors note instead.

> ## Challenge: Investigate the waterway lines

> > Now we will take a deeper look at the Dutch waterway lines: `waterways_nl`. Let's visualize it with the `plot` function. Can you tell what is wrong with this vector file?

Author: Good point. I added some descriptions.

This will write it to disk (in this case, in 'shapefile' format), containing only the data from our cropped area. It can be read again at a later time using the `read_file()` method we have been using above. Note that this actually writes multiple files to disk (`fields_cropped.cpg`, `fields_cropped.dbf`, `fields_cropped.prj`, `fields_cropped.shp`, `fields_cropped.shx`). All these files should ideally be present in order to re-read the dataset later, although only the `.shp`, `.shx`, and `.dbf` files are mandatory (see the [Introduction to Vector Data]({{site.baseurl}}/02-intro-to-vector-data) lesson for more information.)

## Selecting spatial features

From now on, we will take in a point dataset `brogmwvolledigeset.zip`, which is the underground water monitoring wells. We will perform vector processing on this dataset, together with the cropped field polygons `fields_cropped.shp`.

Let's read the two datasets.

~~~
fields = gpd.read_file("fields_cropped.shp")
wells = gpd.read_file("data/brogmwvolledigeset.zip")
~~~
{: .language-python}

And take a look at the wells:

~~~
wells.plot(markersize=0.1)
~~~
{: .language-python}
![All wells](../fig/E07-03-wells-nl.png)

> The points represents all the wells over the Netherlands. Since the wells are in the lat/lon coordinates. To make it comparable with fields, we need to transfer the CRS to RD first:

Reviewer: ```suggestion
The points represents all the wells over the Netherlands. Since the wells are in the lat/lon coordinates. To make it comparable with fields, we need to first transfer the CRS to the "RD New" projection:
```
